File: python/programmers/월간코드챌린지시즌2/모두0으로만들기.py
from collections import deque
def solution(a, edges):
    n = len(a)
    adj=[[] for i in range(n)]
    dp = [0]*n
    visited = [0]*n
    for x,y in edges:
        adj[x].append(y)
        adj[y].append(x)
        dp[x]+=1
        dp[y]+=1
    q = deque()
    for i in range(n):
        if dp[i]==1:
            q.append(i)
    res = 0
    for i in range(n-1):
        x=q.popleft()
        visited[x]=1
        for nx in adj[x]:
            dp[nx]-=1
            if visited[nx]==0:
                res+=abs(a[x])
                a[nx]+=a[x]
                a[x]-=a[x]
            if dp[nx]==1 and visited[nx]==0:
                q.append(nx)
    if sum(a)==0:
        return res
    else:
        return -1
        
print(solution([-5,0,2,1,2],[[0,1],[3,4],[2,3],[0,3]]))     

File: javascript/programmers/Level3/모두0으로만들기.js
function solution(a, edges) {
  const sum = a.reduce((acc, cur) => acc + cur, 0);
  if (sum !== 0) return -1;
  const n = a.length;
  const visited = Array(n).fill(false);
  const arr = Array.from(Array(n), () => []);
  const q = [[0, null]];
  let answer = 0n;

  edges.forEach(([x, y]) => {
    arr[x].push(y);
    arr[y].push(x);
  });

  while (q.length) {
    const [x, y] = q.pop();
    if (visited[x]) {
      answer += BigInt(Math.abs(a[x]));
      a[y] += a[x];
      a[x] = 0;
      console.log(x, y, a);
      continue;
    }
    visited[x] = true;
    q.push([x, y]);
    arr[x].forEach((nx) => {
      if (!visited[nx]) {
        q.push([nx, x]);
      }
    });
  }
  return answer;
}
